executor/aggfuncs/func_count.go

```diff
@@ -0,0 +1,377 @@
+	p := (*partialResult4CountWithDistinct)(pr)
+	p.count = 0
+	p.valSet = newStringSet()
+}
+
+func (e *countOriginalWithDistinct) AppendFinalResult2Chunk(sctx sessionctx.Context, pr PartialResult, chk *chunk.Chunk) error {
+	p := (*partialResult4CountWithDistinct)(pr)
+	chk.AppendInt64(e.ordinal, p.count)
+	return nil
+}
+
+func (e *countOriginalWithDistinct) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) (err error) {
+	p := (*partialResult4CountWithDistinct)(pr)
+	buf := []byte{}
+
+	for _, row := range rowsInGroup {
+		encodedBytes := make([]byte, 0)
+		hasNull, isNull := false, false
+
+		for i := 0; i < len(e.args) && !hasNull; i++ {
+			encodedBytes, isNull, err = e.evalAndEncode(sctx, e.args[i], row, buf, encodedBytes)
+			if err != nil {
+				return
+			}
+			if isNull {
+				hasNull = true
+				break
+			}
+		}
+		if hasNull || p.valSet.exist(string(encodedBytes)) {
+			continue
+		}
+		p.valSet.insert(string(encodedBytes))
+		p.count++
+	}
+
+	return nil
+}
+
+func (e *countOriginalWithDistinct) evalAndEncode(
+	sctx sessionctx.Context, arg expression.Expression,
+	row chunk.Row, buf, encodedBytes []byte,
+) ([]byte, bool, error) {
+	switch tp := arg.GetType().EvalType(); tp {
+	case types.ETInt:
+		val, isNull, err := arg.EvalInt(sctx, row)
+		if err != nil || isNull {
+			return encodedBytes, isNull, errors.Trace(err)
+		}
+		encodedBytes = appendInt64(encodedBytes, val)
+	case types.ETReal:
+		val, isNull, err := arg.EvalReal(sctx, row)
+		if err != nil || isNull {
+			return encodedBytes, isNull, errors.Trace(err)
+		}
+		encodedBytes = appendFloat64(encodedBytes, val)
+	case types.ETDecimal:
+		val, isNull, err := arg.EvalDecimal(sctx, row)
+		if err != nil || isNull {
+			return encodedBytes, isNull, errors.Trace(err)
+		}
+		encodedBytes = appendDecimal(encodedBytes, val)
+	case types.ETTimestamp, types.ETDatetime:
+		val, isNull, err := arg.EvalTime(sctx, row)
+		if err != nil || isNull {
+			return encodedBytes, isNull, errors.Trace(err)
+		}
+		encodedBytes = appendTime(encodedBytes, val)
+	case types.ETDuration:
+		val, isNull, err := arg.EvalDuration(sctx, row)
+		if err != nil || isNull {
+			return encodedBytes, isNull, errors.Trace(err)
+		}
+		encodedBytes = appendDuration(encodedBytes, val)
+	case types.ETJson:
+		val, isNull, err := arg.EvalJSON(sctx, row)
+		if err != nil || isNull {
+			return encodedBytes, isNull, errors.Trace(err)
+		}
+		encodedBytes = appendJSON(encodedBytes, val)
+	case types.ETString:
+		val, isNull, err := arg.EvalString(sctx, row)
+		if err != nil || isNull {
+			return encodedBytes, isNull, errors.Trace(err)
+		}
+		encodedBytes = appendString(encodedBytes, val)
+	default:
+		return nil, false, errors.Errorf("unsupported column type for encode %d", tp)
+	}
+	return encodedBytes, false, nil
+}
+
+func appendInt64(encodedBytes []byte, val int64) (_ []byte) {
+	buf := make([]byte, 8)
```

Comment:
we can pass `buf` as a parameter of this function and reuse it if possible to avoid frequently object allocation.